### ISSUE TITLE
Error message in async fn refers to lifetime `'1` without defining it

### ISSUE THREAD

**Issue by SNCPlay42** [rust-lang/rust#74072]
```rust
pub async fn f(x: &mut i32) -> &i32 {
    let y = &*x;
    *x += 1;
    y
}
```

Output on 1.46.0-nightly (2020-07-04 0cd7ff7ddfb75a38dca8) ([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=245e0d242a85eb60ddeff344d5e55cda))

```
   Compiling playground v0.0.1 (/playground)
error[E0506]: cannot assign to `*x` because it is borrowed
 --> src/lib.rs:3:5
  |
2 |     let y = &*x;
  |             --- borrow of `*x` occurs here
3 |     *x += 1;
  |     ^^^^^^^ assignment to borrowed `*x` occurs here
4 |     y
  |     - returning this value requires that `*x` is borrowed for `'1`
5 | }
  | - return type of generator is &'1 i32

error: aborting due to previous error

For more information about this error, try `rustc --explain E0506`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

Without `async`, a note like this is emitted, and should also be emitted with `async`:

```
1 | pub fn f(x: &mut i32) -> &i32 {
  |             - let's call the lifetime of this reference `'1`
```

Output on stable 1.44.1 (with async) is different:

```
   Compiling playground v0.0.1 (/playground)
error[E0506]: cannot assign to `*x` because it is borrowed
 --> src/lib.rs:3:5
  |
2 |     let y = &*x;
  |             --- borrow of `*x` occurs here
3 |     *x += 1;
  |     ^^^^^^^ assignment to borrowed `*x` occurs here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0506`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by SNCPlay42** [comment]
This note

```
5 | }
  | - return type of generator is &'1 i32
```

Is the one that is supposed to define `'1` (it's produced by [`highlight_region_name`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/borrow_check/diagnostics/region_name/struct.RegionName.html#method.highlight_region_name)), but it's not clear that it's doing so - it appears after the mention of `'1`, not before, and has a poor span.

**Comment by tmandry** [comment]
Discussed in @rust-lang/wg-async-foundations. We think rewording the message to something like "let's call the return type of the async function `&'1 i32`" is a good way to improve this message. But further improvements will be needed to bring it up to par with the non-async version.

**Comment by nikomatsakis** [comment]
In particular, we would *really* like to cite the function *parameter*, not the return type, which is kind of random. But I guess I'd have to dig into the code to figure out how we can make that association.

**Comment by SNCPlay42** [comment]
@rustbot claim

I've got as far as

```
error[E0506]: cannot assign to `*x` because it is borrowed
  --> issue-74072.rs:5:5
   |
LL | pub async fn f(x: &mut i32) -> &i32 {
   |                                ---- return type of async function is &'1 i32
LL |     let y = &*x;
   |             --- borrow of `*x` occurs here
LL |     *x += 1;
   |     ^^^^^^^ assignment to borrowed `*x` occurs here
LL |     y
   |     - returning this value requires that `*x` is borrowed for `'1`
```

I think I can make further improvements.

**Comment by jyn514** [comment]
@SNCPlay42 are you interested in following up on https://github.com/rust-lang/rust/issues/74072#issuecomment-661942271? I don't see a linked PR with those changes, but even if they're not perfect it would be great to have something :)

**Comment by SNCPlay42** [comment]
@jyn514 sorry for the wait! PR is up

**Comment by tmandry** [comment]
@SNCPlay42 Should this be closed or is there more work to do?

For now,

@rustbot release-assignment

Feel free to claim again if you plan to do more work on this.

**Comment by SNCPlay42** [comment]
It looks like #78164 will fix this completely.

There's still room for improvement around closures (they still highlight the closing brace of the block, and it would be nice to use consistent "let's call this..." language in all cases) but these might be appropriate for separate issues (they also aren't specific to async closures).

**Comment by tmandry** [comment]
Triage: looks fixed (probably by #78164). Output on nightly today:

```
error[E0506]: cannot assign to `*x` because it is borrowed
 --> src/lib.rs:3:5
  |
1 | pub async fn f(x: &mut i32) -> &i32 {
  |                   - let's call the lifetime of this reference `'1`
2 |     let y = &*x;
  |             --- borrow of `*x` occurs here
3 |     *x += 1;
  |     ^^^^^^^ assignment to borrowed `*x` occurs here
4 |     y
  |     - returning this value requires that `*x` is borrowed for `'1`
```
